src/util/fs.js

```diff
@@ -251,6 +251,179 @@ async function buildActionsForCopy(
   }
 }
 
+async function buildActionsForHardlink(
+  queue: CopyQueue,
+  events: CopyOptions,
+  possibleExtraneous: Set<string>,
+  reporter: Reporter,
+): Promise<CopyActions> {
+  const artifactFiles: Set<string> = new Set(events.artifactFiles || []);
+  const files: Set<string> = new Set();
+
+  // initialise events
+  for (const item of queue) {
+    const onDone = item.onDone;
+    item.onDone = () => {
+      events.onProgress(item.dest);
+      if (onDone) {
+        onDone();
+      }
+    };
+  }
+  events.onStart(queue.length);
+
+  // start building actions
+  const actions: CopyActions = [];
+
+  // custom concurrency logic as we're always executing stacks of 4 queue items
+  // at a time due to the requirement to push items onto the queue
+  while (queue.length) {
+    const items = queue.splice(0, 4);
+    await Promise.all(items.map(build));
+  }
+
+  // simulate the existence of some files to prevent considering them extraenous
+  for (const file of artifactFiles) {
+    if (possibleExtraneous.has(file)) {
+      reporter.verbose(reporter.lang('verboseFilePhantomExtraneous', file));
+      possibleExtraneous.delete(file);
+    }
+  }
+
+  for (const loc of possibleExtraneous) {
+    if (files.has(loc)) {
+      possibleExtraneous.delete(loc);
+    }
+  }
+
+  return actions;
+
+  //
+  async function build(data): Promise<void> {
+    const {src, dest} = data;
+    const onFresh = data.onFresh || noop;
+    const onDone = data.onDone || noop;
+    files.add(dest);
+
+    if (events.ignoreBasenames.indexOf(path.basename(src)) >= 0) {
+      // ignored file
+      return;
+    }
+
+    const srcStat = await lstat(src);
+    let srcFiles;
+
+    if (srcStat.isDirectory()) {
+      srcFiles = await readdir(src);
+    }
+
+    const destExists = await exists(dest);
+    if (destExists) {
+      const destStat = await lstat(dest);
+
+      const bothSymlinks = srcStat.isSymbolicLink() && destStat.isSymbolicLink();
+      const bothFolders = srcStat.isDirectory() && destStat.isDirectory();
+      const bothFiles = srcStat.isFile() && destStat.isFile();
+
+      if (srcStat.mode !== destStat.mode) {
+        try {
+          await access(dest, srcStat.mode);
+        } catch (err) {
+          // EINVAL access errors sometimes happen which shouldn't because node shouldn't be giving
+          // us modes that aren't valid. investigate this, it's generally safe to proceed.
+          reporter.verbose(err);
+        }
+      }
+
+      // correct hardlink
+      if (bothFiles && (srcStat.ino === destStat.ino)) {
```

Comment:
Indeed there were some issues with ino on Windows back in Node 0.10 and 0.12 times https://github.com/nodejs/node-v0.x-archive/issues/2670.
The tests on Appveyor pass though.
I'll google a bit more to double check if we can rely on ino not being null